add a code of conduct

ashleygwilliams opened the pull request:
this is npm's coc with some minor changes.
the node code of conduct is not complete enough.
i would like to enforce a more specific and actionable one here.
if node is interested in upstreaming this, that would be wonderful.

would love help cleaning out the npm specific content.
also we need an email address for abuse reports.

if you find mistakes please feel free to submit a PR to this branch! let me know if you need help with that at all.

Trott commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,126 @@
+
+* Be respectful.
+* Abusive behavior is never tolerated.
+  administrators, and may be removed.
```
Copy/paste error?

ashleygwilliams replied:
lol yup. i knew i'd miss a bunch on the first pass. thanks!

thefourtheye commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+
+## Scope
+
+Wea expect all members and participants of the Node.js Inclusivity
```
We?

ashleygwilliams replied:
hehe thanks!

thefourtheye commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+
+## Friendly Harassment-Free Space
+
+We are committed to providing a friendly, safe and welcoming
```
`committed to provide`?

nebrius replied:
technically, `committed to providing` is the correct grammatical phrasing (English is a weird language, why did it have to become the default?)

thefourtheye replied:
@nebrius :open_mouth: I always thought, `to` should not be followed by continuous verb.

nebrius replied:
@thefourtheye that's why English is a weird language, Every single grammatical rule in English should be read as "Do things this way, except in special cases 1, 2, 3, 4, 5, 6, ..., 543,765.

The number of edge cases in English grammar where we do the opposite of the rule is horrifying, and I have a _lot_ of respect for all of you who don't speak it as your 1st language.

thefourtheye replied:
@nebrius Well, thanks for bearing with us and helping us learn :-)

Trott replied:
In this particular case, `providing` is a gerund. It functions as a noun, not a verb. Other examples of `to` being followed by a gerund rather than being part of an infinitive verb would be `addicted to gambling` and `averse to smoking`.

thefourtheye replied:
Ah that makes sense now. Gotta understand the Grammar better now. Thanks @Trott :-)

nebrius replied:
TIL. I didn't even know what a gerund is, thanks @Trott :)

thefourtheye commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+identifiers that might detract from a friendly, safe, and welcoming
+environment for all.
+
+Harassment includes, but is not limited to: harmful or prejudicial
```
This part is repeating from few paragraphs above.

ashleygwilliams replied:
it's true. repetition is good in this case IMO

scottgonzalez replied:
Perhaps rephrasing in more general terms will avoid the duplication while still repeating the intention.

Something along the lines of "Harassment includes all forms of verbal, written, and physical actions which may cause emotional, mental, or physical harm to another person."

scottgonzalez replied:
If we go down this path and the duplication is removed, the remaining portion after the semicolon should still be kept as it's not redundant.

ashleygwilliams replied:
i think i'd like to keep it. i did s/disability/ability tho as per another one of your comments @scottgonzalez

thefourtheye commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+If you believe someone is harassing you or has otherwise violated this
+Code of Conduct, please contact us at <??????????> to send us an
+abuse report.  If this is the initial report of a problem, please
+include as much detail as possible. It is easiest for us to address
```
`It is easy for us to`?

jasnell replied:
"easiest" is used correctly here. The sentence is still a bit awkward tho.

ashleygwilliams replied:
i am going to keep this unless someone else has a suggestion.

scottgonzalez replied:
"The more context we have, the better we can assess and resolve the situation"?

ashleygwilliams replied:
+1 @scottgonzalez

thefourtheye commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+
+This Code of Conduct borrows heavily from the Stumptown Syndicate
+[Citizen's Code of Conduct](http://citizencodeofconduct.org/), and the
+[Rust Project Code of
```
Inconsistent line breakages... Looks like editor automatically did them :D

ashleygwilliams replied:
yar, i'll go back and fix that

ashleygwilliams replied:
haha editor is `vi`

jasnell replied:
Out of curiosity, why start with npm's code of conduct when Node.js already has one? Why not start off with that basis and extend with the additional pieces you feel are necessary?

nebrius replied:
I can't speak for Ashley, but I would claim that using the NPM CoC is much more expedient. The Node.js CoC would require a lot of changes to get it where we want it to be for this WG.

Writing a CoC is really hard work, as it tends to skirt next to legal issues in some cases. I like to think of CoC's as being analogous to crypto software. The first rule of crypto software is "don't write your own crypto software" :)

Same thing with CoC's, "don't write your own CoC" is a very good rule to live by.

ashleygwilliams replied:
yup @nebrius. @jasnell there are no enforcement procedures documented in the current code of conduct. also the current code of conduct lacks the rigor to be enforcable. the npmjs one has both (they are otherwise very similar!).

mikeal replied:
IMO, this CoC already uses simpler language than the Rust CoC we have in Core which makes it easier for translating and non-native english speakers. I'm sure there are improvements we can continue to make but more people in this WG were involved in drafting the npm CoC than the Rust CoC so it makes sense to start with this as a base.

nebrius commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+
+The definitions of various subjective terms such as "discriminatory",
+"hateful", or "confusing" will be decided at the sole discretion of
+the Node.js Inclusivity Working Group team members.
```
nitpick, can we rename `team members` to just `members`? I think that the term `members` is used more frequently.

sup commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+> Node.js Inclusivity WG
```
perhaps add a note that this is an altered version of the npm coc to clear up potential confusion?

nebrius replied:
+1 to mentioning that this is based on NPM's. Perhaps put it in the `Credit and License` section?

ashleygwilliams replied:
yup good call.

jasnell commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+* Be respectful.
+* Abusive behavior is never tolerated.
+* Violations of this code may result in swift and permanent expulsion
+  from the Node.js inclusivity working group and/or the Node.js
```
If this is the CoC for the Inclusivity WG, then the part about expulsion from the Node.js community should be stricken

nebrius replied:
What about cases where they also violate the project wide CoC? I think that's the intention with this line, and in that case, they should be banned from the Node.js community as a whole.

jasnell replied:
"they" being individual TSC members? The expectation would be then that a report would be made to the full TSC, which would then determine the appropriate course of action. The individual reported would be expected to recuse themselves from that discussion.

nebrius replied:
"they" being the offender. I would claim that this should be sent to the not yet formed Moderation WG (see https://github.com/nodejs/inclusivity/issues/16), not the TSC. We don't want to bog down the TSC with having to deal with these situations unless a special case arises, just like we don't want to bog down the inclusivity WG either.

Most bans will be really uncontroversial, and involving the TSC isn't a good use of their time. Just look at the 4chan/8chan harassment we had on Friday. That's not really something TSC members will want to do, because honestly it's not work that _anyone_ wants to do ;-)

jasnell replied:
For most reports I agree absolutely. For reports involving TSC members, those have to go to the TSC. It's part of the board-approved charter that only the TSC can sanction/remove one of it's own members so the language/process definitely needs to be clear on that. The moderation WG can be chartered to handle that responsibility but there's no proposed moderation WG charter we can point at currently.

ashleygwilliams replied:
please note the "may" and the "and/or"

it is possible. it is not deterministic. this language captures that just fine IMO.

mikeal replied:
> For reports involving TSC members, those have to go to the TSC. It's part of the board-approved charter that only the TSC can sanction/remove one of it's own members so the language/process 

This all depends on how this gets setup.

It's best to think about it this way: all rights and responsibilities are first owned by the board, then they are passed to the TSC, then the TSC can pass them to top level projects and WGs. The TSC charter passed almost all project related responsibilities to the TSC (from the board) with only a few exceptions like legal issues. If some project related responsibility happened to be left out of the charter the board would more than likely amend it to include it.

The TSC can move **any** of its rights and responsibilities to another group, there's no restriction on that. When the TSC created the Core TLP it moved about 80% of its charter to that group.

Additionally, every working group charter takes responsibility for its own membership rules. If a working group wants to exclude or moderate someone, even a member of the TSC, they have that right. Once responsibilities are passed to a WG the only recourse the TSC has is to revoke the entire charter, they don't reserve the right to pick at individual issues without being members of that group. So, moderation and membership *within the inclusivity WG* is at the discretion of this WG, not the TSC, once the charter is ratified if we're going by what has been the standard of every WG ratified so far.

*If* in the future a WG creates a charter that asks the TSC to hand over foundation-wide, rather than only WG-wide, conduct responsibilities the TSC *can* approve that charter and, unless explicitly excluded, that would include the conduct of TSC members. That isn't happening yet, but it's worth nothing that it is possible. That's why charters tend to be picked over pretty heavily before they are ratified and responsibilities tend to be very explicit to avoid granting unintended rights -- which is one of the reasons you keep seeing @jasnell ask that WG scope be explicit in so many places.

jasnell replied:
@mikeal ... yep. it may seem like I'm nitpicking, but the importance of precision of language and intent cannot be understated.

ashleygwilliams replied:
what if i wrote "may be escalated to the TSC moderation team which could result in expulsion from the nodejs community?"

i def agree with everyone's thoughts on scope of this wg but do want to indicate this escalation path, which, in certain cases, would be a likely trajectory

jasnell commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+
+If a community member engages in unacceptable behavior, the Working
+Group members may take any action they deem appropriate, up to and
+including a temporary ban or permanent expulsion from the community
```
Again, this should be scoped to participation in this WG's activities. Unless specifically chartered to do so by the TSC, this WG cannot expel individuals from the community (unless by "community" you are referring to the Inclusivity WG, in which case this needs to be clarified)

nebrius replied:
Perhaps something like `the Working Group members...from this Working Group without warning, and, at their discretion, may recommend expulsion from the Node.js community as a whole`?

nebrius replied:
Once again, I want to emphasize that we _do not_ want the ability to have enforecement powers outside of this WG. I ask that you please give us the benefit of the doubt on this, and assume that if that's the implication, it's the result of poor wording, not malicious intent.

jasnell replied:
+1. On the TSC side, we should have a process in place requiring that such reports be handled within a specific period of time. @mikeal, we should make sure that's on the next TSC agenda.

jasnell replied:
@nebrius ... no malicious intent is assumed. It's just important to be absolutely clear on the wording to avoid misunderstandings later on.

nebrius replied:
I'm sorry if I misinterpreted your comment. We've been told many many times about scope of enforcement powers, and have also told others many many times about scope of enforcement powers. At this point, further explaining why we need be careful about scope is not necessary. Trust me, _we know_ ;-)

jasnell replied:
Personally I find it very important to be as precise as possible with the language on these kinds of things and I will continue to point out places where I feel the language needs to be clarified. I would expect (and have expected) no less were I the one writing/editing the documents.

ashleygwilliams replied:
this code of conduct and it's being enforced is exclusively in the scope of the working group. not the rest of node. this is explicitly stated at the beginning.

jasnell commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+## Addressing Grievances
+
+If you feel you have been falsely or unfairly accused of violating
+this Code of Conduct, you should notify the Working Group.  We will
```
Notify how? If the person has been banned from participation in any of the discussion forums, they will have no means of communicating with the WG. So a channel of communication must be provided and documented.

nebrius replied:
See `Contact Info` below

jasnell replied:
:-) saw that after I left the comment. I was working top to bottom. A forward link may be helpful.

nebrius replied:
Good point, +1 to a forward link

ashleygwilliams replied:
:+1:

jasnell commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+
+If you feel you have been falsely or unfairly accused of violating
+this Code of Conduct, you should notify the Working Group.  We will
+do our best to ensure that your grievance is handled appropriately.
```
Handled how? The appeal process should be documented.

ashleygwilliams replied:
it should be but this is a good start.

mikeal commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+the Working Group members.
+
+Unacceptable behavior from any community member, including sponsors,
+employees, customers, or others with decision-making authority, will
```
How translatable is "decision-making authority"? I'm not an expert on this but if I had to use this word when traveling in a foreign country I know I'd find it difficult so I'm wondering if there is simpler language we can use instead.

sup replied:
now that you mention it, i wouldn't be able to translate it into german without paraphrasing it in some way. how about `others with the ability to make decisions` (although that sounds _too_ simple)

sup replied:
another suggestion: `others with controlling/managing abilities`

thefourtheye replied:
Simply put, TSC.

ashleygwilliams replied:
good call! thanks for this i didn't consider it.

mikeal replied:
I know that I started this whole thread but I'm realizing that picking at the language from this point of view might be better accomplished by first landing the CoC and then asking a few people to begin translating while sending PRs to adjust any language that is difficult to translate. The end result should be simpler english that also helps with people finding the english version that aren't experts in english.

ashleygwilliams replied:
hehe yes. i'm gonna wait for a few more hours and then try to consolidate all the suggestions and spit out a new version

jasnell commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+abuse report.  If this is the initial report of a problem, please
+include as much detail as possible. It is easiest for us to address
+issues when we have more context.
+
```
Would recommend some indication that the details of the report, including the identity of both parties, will be kept confidential and shared only on a need to know basis *unless* a determination is made that the reported individual should be recommended for banning. 

Specifically, while a matter is being investigated, members of this WG should not be sharing details of any reported abuse in any forum other than those established by this WG specifically for that purpose.

nebrius replied:
+1. I would expand on this and add that some parts of the report, specifically the names of those who were harassed, should be kept confidential indefinitely.

ashleygwilliams replied:
seems good to me.

jasnell commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+This policy covers behavior in the context of the Node.js Inclusivity
+Working Group Open Source community, including but not limited to public
+GitHub repositories, and IRC/Slack channels.
+
```
Given this [comment](https://github.com/nodejs/inclusivity/pull/22/files#r44876649), recommend adjusting the two paragraphs above to read, "*This policy covers behavior within the context of the Node.js Inclusivity Working Group. It is expected that all Working Group members, Node.js community members, and commenters participating in Inclusivity Working Group affairs abide by this Code of Conduct at all times in all public or private venues.*"

The intent is only to make the language a bit more precise.

ashleygwilliams replied:
i think this is good! updating with a few little tweaks.

bnoordhuis commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+address a grievance related to an abuse report.
+
+You are also encouraged to contact us if you are curious about
+something that might be "on the line" between appropriate and
```
Potentially confusing to non-native speakers.  What about "if you are unsure if something is appropriate or inappropriate"?

sup replied:
+1

ashleygwilliams replied:
:+1:

bnoordhuis commented:
I don't like the repeated use of "the Node.js community" in this document.  Community covers a lot more than what's under https://github.com/nodejs.

ashleygwilliams commented:
@bnoordhuis do you have a suggestion on what you'd like to replace it with? thanks!

bnoordhuis commented:
Personally, I'd just drop it.  If you want to have something there, what about "other Working Groups"?

mikeal commented:
> RE: use of Node.js Commmunity.

I see only two references to "Node.js community" and each is used quite differently. 

This first is regarding banning/expulsion from the "Node.js Community." I don't know entirely what that means but I would guess that it means repos, orgs, IRC, and related community events. We could spell those all out but note that moderation above the WG would have to be escalated. While this moderation falls outside *just* the WG it is accurate in that if someone were to post in this repo similar anti-semitic images like we saw yesterday in the TSC repo they would be banned from all community related spaces (if they could be identified and aren't using a throwaway account).

The second reference is "Node.js community venues, online and in person" in regard to being held accountable in this repo/WG for conduct outside this repo/WG (like a conference or twitter). Basically, if someone is a known harasser in another space they won't be allowed to participate in this one. For this reference we could probably just remove "Node.js community" and reference "all venues online and in-person" as I don't think we would *exclude* serious conduct violations that weren't explicitly Node.js venues.

bnoordhuis commented:
To clarify, when I read the word "community", I think of more than just this project.  Express, hapi, my local user group, those are all part of the wider community, and outside our remit.

jasnell commented:
The "I don't know entirely what that means" is the challenge,
unfortunately.
On Nov 15, 2015 1:53 PM, "Mikeal Rogers" <notifications@github.com> wrote:

> RE: use of Node.js Commmunity.
>
> I see only two references to "Node.js community" and each is used quite
> differently.
>
> This first is regarding banning/expulsion from the "Node.js Community." I
> don't know entirely what that means but I would guess that it means repos,
> orgs, IRC, and related community events. We could spell those all out but
> note that moderation above the WG would have to be escalated. While this
> moderation falls outside *just* the WG it is accurate in that if someone
> were to post in this repo similar anti-semitic images like we saw yesterday
> in the TSC repo they would be banned from all community related spaces (if
> they could be identified and aren't using a throwaway account).
>
> The second reference is "Node.js community venues, online and in person"
> in regard to being held accountable in this repo/WG for conduct outside
> this repo/WG (like a conference or twitter). Basically, if someone is a
> known harasser in another space they won't be allowed to participate in
> this one. For this reference we could probably just remove "Node.js
> community" and reference "all venues online and in-person" as I don't think
> we would *exclude* serious conduct violations that weren't explicitly
> Node.js venues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/inclusivity/pull/22#issuecomment-156858851>.
>

scottgonzalez commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+Working Group and community, including commenters and contributors, to
+abide by this Code of Conduct at all times in all Node.js community
+venues, online and in person, and in one-on-one communications
+pertaining to the Node.js Working Group's  affairs.
```
double space before the last word

scottgonzalez commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+
+We are committed to providing a friendly, safe and welcoming
+environment for all, regardless of gender identity, sexual
+orientation, disability, ethnicity, religion, age, physical
```
Disability is fairly vague and also seems to be the only word in this list that starts with a negative connotation. Perhaps something along the lines of "physical or mental ability"? Generally with disabilities, it's better to refer to specific disabilities, but that doesn't work with this type of list.

ashleygwilliams replied:
this is a great comment. i might even upstream this to the npm original! thanks!

ashleygwilliams replied:
https://github.com/npm/policies/pull/33

scottgonzalez commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+## Reporting Violations of this Code of Conduct
+
+If you believe someone is harassing you or has otherwise violated this
+Code of Conduct, please contact us at <??????????> to send us an
```
Just marking this line to make sure the contact info is filled in later.

scottgonzalez commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,124 @@
+
+## Contact Info
+
+Please contact <?????????> if you need to report a problem or
```
Again just marking so this doesn't get missed later.

Fishrock123 replied:
@nodejs/inclusivity Should we get this before merging?

ashleygwilliams commented:
@bnoordhuis @jasnell @mikeal i believe all references to the Node.js community have been removed except for one place where it refers specfically to Node.js community members particupating in this working group (actually a direct contribution of wording from @jasnell).

is this better? thoughts on the current draft now that i've addressed all the comments?

ashleygwilliams commented:
ok waiting on https://github.com/nodejs/email/issues/9 and then merging

beaugunderson commented:
@ashleygwilliams read through all of the discussion and I think everything has been addressed :+1: :sparkles:

nebrius commented:
LGTM, just one last question. Did we actually get the email `inclusivity@nodejs.com`? I _think_ we only got `inclusivity@iojs.org`: https://github.com/nodejs/email/commit/ea9a396c2c54e3ac2ed566c536f723d4e5810e59

ashleygwilliams commented:
the issue made it seem like they had contacted the people and got it working. lemme check my email. i got a test message.

beaugunderson commented:
I do recall that the @nodejs.org alias was added as well. :)

nebrius commented:
LGTM! I'll leave to @beaugunderson to +1 and merge.

beaugunderson commented:
:+1:!
